realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp

```diff
@@ -27,135 +27,147 @@ static void finalize_unchecked_row(jlong ptr);
+JNIEXPORT jobjectArray JNICALL Java_io_realm_internal_UncheckedRow_nativeGetColumnNames(JNIEnv* env, jobject,
+                                                                               jlong nativeRowPtr)
 {
-    TR_ENTER_PTR(nativeRowPtr)
-    if (!ROW(nativeRowPtr)->is_attached()) {
-        return 0;
+    if (!OBJ(nativeRowPtr)->is_valid()) {
+        ThrowException(env, IllegalArgument, "Object passed is not valid");
     }
 
     try {
-        JStringAccessor columnName2(env, columnName);                                   // throws
-        return to_jlong_or_not_found(ROW(nativeRowPtr)->get_column_index(columnName2)); // noexcept
+        ColKeys col_keys = OBJ(nativeRowPtr)->get_table()->get_column_keys();
+
+        size_t size = col_keys.size();
+        jobjectArray col_keys_array = env->NewObjectArray(size, JavaClassGlobalDef::java_lang_string(), 0);
+        if (col_keys_array == NULL) {
+            ThrowException(env, OutOfMemory, "Could not allocate memory to return column keys.");
+            return NULL;
+        }
+        for (size_t i = 0; i < size; ++i) {
+            env->SetObjectArrayElement(col_keys_array, i, to_jstring(env,  OBJ(nativeRowPtr)->get_table()->get_column_name(col_keys[i])));
+        }
+
+        return col_keys_array;
+
     }
     CATCH_STD()
-    return 0;
+    return NULL;
 }
 
+
 JNIEXPORT jint JNICALL Java_io_realm_internal_UncheckedRow_nativeGetColumnType(JNIEnv*, jobject, jlong nativeRowPtr,
-                                                                               jlong columnIndex)
+                                                                               jlong columnKey)
 {
-    TR_ENTER_PTR(nativeRowPtr)
-    auto column_type = ROW(nativeRowPtr)->get_column_type(S(columnIndex)); // noexcept
-    if (column_type != type_Table) {
-        return static_cast<jint>(column_type);
-    }
-    // FIXME: Add test in https://github.com/realm/realm-java/pull/5221 before merging to master
-    return static_cast<jint>(ROW(nativeRowPtr)->get_table()->get_descriptor()->get_subdescriptor(S(columnIndex))->get_column_type(S(0))
-                             + io_realm_internal_Property_TYPE_ARRAY); // noexcept
+    ColKey column_key (columnKey);
+    auto table = OBJ(nativeRowPtr)->get_table();
+    jint column_type = table->get_column_type(column_key);
+    if (table->is_list(column_key) && column_type < type_LinkList) {
+        // add the offset so it can be mapped correctly in Java (RealmFieldType#fromNativeValue)
+        column_type += 128;
+    }
+
+    return column_type;
 }
 
-JNIEXPORT jlong JNICALL Java_io_realm_internal_UncheckedRow_nativeGetIndex(JNIEnv* env, jobject, jlong nativeRowPtr)
+JNIEXPORT jlong JNICALL Java_io_realm_internal_UncheckedRow_nativeGetObjectKey(JNIEnv* env, jobject, jlong nativeRowPtr)
 {
-    TR_ENTER_PTR(nativeRowPtr)
-    if (!ROW_VALID(env, ROW(nativeRowPtr))) {
+    if (!ROW_VALID(env, OBJ(nativeRowPtr))) {
         return 0;
```

Comment:
Seems weird. The invalid ObjKey is generated by ObjKey(). So getting the internal value of an invalid ObjKey should be ObjKey().value instead of just '0'

Reply:
the macro throws an exception btw so the return `0` is for the compiler…